### ISSUE TITLE
Build failure on GHC 8.x (macOS 10.14): ld: unknown option: -no_fixup_chains

### ISSUE THREAD

**Issue by andreasabel** [haskell/zlib#53]
I am getting build errors on macOS 10.14 which I have not seen before.  They happen with any GHC 8 version and disappear with GHC 9:
```
$ cabal build -w ghc-8.10.7
...
Preprocessing library for zlib-0.6.3.0..
linking /Users/abel/bin/src/zlib/dist-newstyle/build/x86_64-osx/ghc-8.10.7/zlib-0.6.3.0/build/Codec/Compression/Zlib/Stream_hsc_make.o failed (exit code 1)
rsp file was: "/Users/abel/bin/src/zlib/dist-newstyle/build/x86_64-osx/ghc-8.10.7/zlib-0.6.3.0/build/Codec/Compression/Zlib/hsc2hscall34067-2.rsp"
command was: /usr/bin/gcc /Users/abel/bin/src/zlib/dist-newstyle/build/x86_64-osx/ghc-8.10.7/zlib-0.6.3.0/build/Codec/Compression/Zlib/Stream_hsc_make.o /Users/abel/bin/src/zlib/dist-newstyle/build/x86_64-osx/ghc-8.10.7/zlib-0.6.3.0/build/Codec/Compression/Zlib/Stream_hsc_utils.o -o /Users/abel/bin/src/zlib/dist-newstyle/build/x86_64-osx/ghc-8.10.7/zlib-0.6.3.0/build/Codec/Compression/Zlib/Stream_hsc_make \
  --target=x86_64-apple-darwin -Wl,-no_fixup_chains -L/usr/local/opt/icu4c/lib -L/usr/local/opt/libxml2/lib -lz -L/usr/local/lib/ghc-8.10.7/bytestring-0.10.12.0 -L/usr/local/lib/ghc-8.10.7/deepseq-1.4.4.0 -L/usr/local/lib/ghc-8.10.7/array-0.5.4.0 -L/usr/local/lib/ghc-8.10.7/base-4.14.3.0 -liconv -L/usr/local/lib/ghc-8.10.7/integer-gmp-1.0.3.0 -L/usr/local/lib/ghc-8.10.7/ghc-prim-0.6.1 -L/usr/local/lib/ghc-8.10.7/rts -lm -ldl
error: ld: unknown option: -no_fixup_chains
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
The latest macOS virtual environment does not have this problem: https://github.com/andreasabel/zlib/actions/runs/4194386615/jobs/7272477723#step:5:23
There is only a warning:
> ld: warning: -undefined dynamic_lookup may not work with chained fixups

I am putting this up here, maybe someone has an idea how to work around this.

**Comment by andreasabel** [comment]
Probably my system was broken.  Fine after upgrading to macOS 12.

**Comment by mengwong** [comment]
I encountered this recently and am minded to believe that your system was not broken.

At some point in the macos 13.4.* sequence, support for `no_fixup_chains` was  dropped.

This led to an explicit test in configure: https://gitlab.haskell.org/ghc/ghc/-/commit/a587499c6dc8a7e7e646e9c5f246903a4cc99e56 

which was backported to 9.2 and appears in 9.2.8-release:
https://gitlab.haskell.org/ghc/ghc/-/commits/ghc-9.2.8-release/

However, stack downloads a ghc from
https://downloads.haskell.org/~ghc/9.2.8/ghc-9.2.8-x86_64-apple-darwin.tar.bz2

This ghc does not contain the test; instead hsc2hs includes -no_fixup_chains by default.

```
┌─[mengwong@solo-8] - [~/tmp/ghc-9.2.8-x86_64-apple-darwin] - [2023-07-07 12:20:43]
└─[130] <> rg no_fixup_chains
wrappers/hsc2hs-ghc-9.2.8
1:HSC2HS_EXTRA="--cflag=--target=x86_64-apple-darwin--lflag=--target=x86_64-apple-darwin --lflag=-Wl,-no_fixup_chains"
```

Being a n00b it is not clear to me how exactly the apple-darwin tarball of 9.2.8 differs from the source release. Presumably it is in a "half-baked" state specialized for the OS. Perhaps this is where the difficulty arises.

I found that a later tarball for 9.4.5 does contain the test in configure, but not for 9.4.1.

**Comment by mengwong** [comment]
For anyone struggling with this: a workaround.

I edited my `hsc2hs` executable to remove the ` --lflag=-Wl,-no_fixup_chains` bit on the end of `HSC2HS_EXTRA`

I had `system-ghc: true` specified so I did this under my `.ghcup` others may need to do that under their `~/.stack/programs/`

**Comment by andreasabel** [comment]
@mengwong Is this something that should be reported to the GHC team?  
However, then it would be better to have a independent reproducer which could be hard to come by.   (E.g. I tried it on the GitHub Action runners but was not successful, see my OP.)

**Comment by mengwong** [comment]
I think it would be better reported to the Stack team specifically, but let's wait for some other poor soul to stumble upon this via Google search, and then they can report that they have independently reproduced the experiment.

**Comment by elaforge** [comment]
I hit this same issue when compiling the `clock` package (probably just the first one that uses hsc2hs) on the latest ghc-9.4.7 from ghcup.  And it's easy to see that the official binary distribution at https://downloads.haskell.org/~ghc/9.4.6/ghc-9.4.6-x86_64-apple-darwin.tar.xz includes `wrappers/hsc2hs-ghc-9.4.6` with `-no_fixup_chains`.

I tested with newer OSX and it does support that flag, though it's not documented. 
 So it's probably just that the binary distribution is built on a newer OS version.  Since 10.13.6 is out of support, it's not a big surprise.  So what you said is essentially correct, but backwards, the flag was added not removed.  The workaround does work though, so thanks for that!  Eventually I'll have to upgrade the OS though.
